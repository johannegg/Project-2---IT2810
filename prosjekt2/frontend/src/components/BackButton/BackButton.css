.back-button {
	padding: 8px;
	color: #d9829f;
	background-color: transparent;
	font-size: 18px;
	border: none;
}