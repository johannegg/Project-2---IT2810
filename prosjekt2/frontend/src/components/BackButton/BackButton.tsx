import React from "react";
import { useNavigate } from "react-router-dom";
import "./BackButton.css";

interface BackButtonProps {
	text?: string;
}

const BackButton: React.FC<BackButtonProps> = ({ text = "Go back" }) => {
	const navigate = useNavigate();

	return (
		<button className="back-button" onClick={() => navigate(-1)}>
			&#10094; {text}
		</button>
	);
};

export default BackButton;
