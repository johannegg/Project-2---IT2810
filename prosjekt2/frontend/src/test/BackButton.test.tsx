import { describe, test, vi, beforeEach, afterEach, expect } from "vitest";
import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import BackButton from "../components/BackButton/BackButton";
import { MemoryRouter } from "react-router-dom"; // Provides router context for testing

// Mock implementation of useNavigate from react-router-dom
vi.mock("react-router-dom", async () => {
	const actual = await vi.importActual("react-router-dom");
	return {
		...actual,
		useNavigate: vi.fn(),
	};
});

// Import mocked useNavigate
import { useNavigate } from "react-router-dom";

describe("BackButton", () => {
	const mockNavigate = vi.fn(); // Mock function for navigation

	// Set up mock before each test
	beforeEach(() => {
		vi.mocked(useNavigate).mockReturnValue(mockNavigate);
	});

	// Clean up after each test
	afterEach(() => {
		cleanup();
		vi.resetAllMocks();
	});

	// Test 1: Verifies that BackButton renders with the default text
	test("renders BackButton with default text", () => {
		render(
			<MemoryRouter>
				<BackButton />
			</MemoryRouter>,
		);

		// Assert that the button is rendered with the default text
		const button = screen.getByRole("button", { name: /Go back/i });
		expect(button).toBeInTheDocument();
	});

	// Test 2: Verifies that BackButton renders with custom text
	test("renders BackButton with custom text", () => {
		const customText = "Back to Home";
		render(
			<MemoryRouter>
				<BackButton text={customText} />
			</MemoryRouter>,
		);

		// Assert that the button is rendered with the custom text
		const button = screen.getByRole("button", { name: new RegExp(customText, "i") });
		expect(button).toBeInTheDocument();
	});

	// Test 3: Verifies that clicking the button triggers navigation
	test("navigates back when clicked", () => {
		render(
			<MemoryRouter>
				<BackButton />
			</MemoryRouter>,
		);

		// Simulate a button click
		const button = screen.getByRole("button", { name: /Go back/i });
		fireEvent.click(button);

		// Assert that navigate(-1) is called
		expect(mockNavigate).toHaveBeenCalledWith(-1);
	});

	// Test 4: Snapshot test for BackButton with default props
	test("matches snapshot with default props", () => {
		const { container } = render(
			<MemoryRouter>
				<BackButton />
			</MemoryRouter>,
		);

		// Assert that the rendered output matches the snapshot
		expect(container).toMatchSnapshot();
	});

	// Test 5: Snapshot test for BackButton with custom text
	test("matches snapshot with custom text", () => {
		const { container } = render(
			<MemoryRouter>
				<BackButton text="Custom Text" />
			</MemoryRouter>,
		);

		// Assert that the rendered output matches the snapshot
		expect(container).toMatchSnapshot();
	});
});
