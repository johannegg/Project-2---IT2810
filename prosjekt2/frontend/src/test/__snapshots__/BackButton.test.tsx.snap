// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BackButton > matches snapshot with custom text 1`] = `
<div>
  <button
    class="back-button"
  >
    ❮ 
    Custom Text
  </button>
</div>
`;

exports[`BackButton > matches snapshot with default props 1`] = `
<div>
  <button
    class="back-button"
  >
    ❮ 
    Go back
  </button>
</div>
`;
